Automattic/wp-calypso
# Stats: build odyssey translations

#### Proposed Changes

P2: pejTkB-8l-p2

The PR adds the ability to build single language file for each language. `apps/odyssey-stats/bin/build-languages.js` is forked from `bin/build-languages.js`. The size of a single language is reduced from ~1M to ~80K max.

The overall process is:

- Build the project, where `build-tools/webpack/generate-chunks-map-plugin` generate a `chunks-map.json` for us. We use the file to merge all the strings together referenced by the built asset, for example `build.min.js`.
- Extract all strings to a pot file `odyssey-strings.pot`
- Then `apps/odyssey-stats/bin/build-languages.js` does the rest - merge all referenced strings together, pick only the necessary translations from the full translations files, and lastly write them to the target language files in `dist/languages`
- The last step is to load the built language files for Odyssey in `apps/odyssey-stats/app.jsx`

Next steps:

1. We'll automate the translation process in a follow up PR by adding the following code to `.teamcity/_self/projects/WPComPlugins.kt`:
```
	buildSteps = {
		bashNodeScript {
			name = "Translate Odyssey Stats"
			scriptContent = """
				cd apps/odyssey-stats

				# generate language files
				yarn translate
			"""
		}
	}
```

2. Another thing is, we may want to bust cache for the translations.


#### Testing Instructions

* `cd apps/odyssey-stats`
* `yarn build`
* `yarn translate`
* Ensure translations generated in `dist/languages`
* `rsync -avzP path/to/apps/odyssey-stats/dist/ sandbox:/path/to/widgets.wp.com/odyssey-stats/v1 --delete`
* Sandbox `widgets.wp.com`
* Open Stats dashboard from a Jetpack site
* Ensure Odyssey Stats loads language files properly and the interface is translated

<img width="877" alt="image" src="https://user-images.githubusercontent.com/1425433/211965780-810f38a4-b349-4ec2-81e4-6dcbe20da136.png">


#### Pre-merge Checklist
